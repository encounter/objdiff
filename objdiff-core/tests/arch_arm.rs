use objdiff_core::{diff, obj};

mod common;

#[test]
#[cfg(feature = "arm")]
fn read_arm() {
    let diff_config = diff::DiffObjConfig { ..Default::default() };
    let obj = obj::read::parse(include_object!("data/arm/LinkStateItem.o"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "_ZN13LinkStateItem12OnStateLeaveEi").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "arm")]
fn read_thumb() {
    let diff_config = diff::DiffObjConfig { ..Default::default() };
    let obj = obj::read::parse(include_object!("data/arm/thumb.o"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj
        .symbols
        .iter()
        .position(|s| s.name == "THUMB_BRANCH_ServerDisplay_UncategorizedMove")
        .unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "arm")]
fn combine_text_sections() {
    let diff_config = diff::DiffObjConfig { combine_text_sections: true, ..Default::default() };
    let obj = obj::read::parse(include_object!("data/arm/enemy300.o"), &diff_config).unwrap();
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "Enemy300Draw").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}
