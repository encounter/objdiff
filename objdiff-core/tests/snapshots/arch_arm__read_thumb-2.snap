---
source: objdiff-core/tests/arch_arm.rs
expression: diff.instruction_rows
---
[
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 0,
                size: 2,
                opcode: 56,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 2,
                size: 2,
                opcode: 74,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 4,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 6,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 8,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 10,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 12,
                size: 2,
                opcode: 66,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 14,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 18,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 20,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    212,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 97,
                branch_idx: 0,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 22,
                size: 2,
                opcode: 38,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 24,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 26,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    48,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 22,
                branch_idx: 1,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 28,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 30,
                size: 2,
                opcode: 26,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 32,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    94,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 44,
                branch_idx: 2,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 34,
                size: 2,
                opcode: 35,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 36,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 38,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 40,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 44,
                size: 2,
                opcode: 7,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 46,
                size: 2,
                opcode: 55,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 48,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    12,
                ],
                branch_idx: 1,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 50,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 52,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 56,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 58,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    212,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 97,
                branch_idx: 0,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 60,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 62,
                size: 2,
                opcode: 33,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 64,
                size: 2,
                opcode: 36,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 66,
                size: 2,
                opcode: 42,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 68,
                size: 2,
                opcode: 44,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 70,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    212,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 97,
                branch_idx: 0,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 72,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 74,
                size: 2,
                opcode: 17,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 76,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 78,
                size: 2,
                opcode: 54,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 80,
                size: 2,
                opcode: 67,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 82,
                size: 2,
                opcode: 36,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 84,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 86,
                size: 2,
                opcode: 7,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 88,
                size: 2,
                opcode: 54,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 90,
                size: 2,
                opcode: 67,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 92,
                size: 2,
                opcode: 55,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 94,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    15,
                ],
                branch_idx: 2,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 96,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 98,
                size: 2,
                opcode: 4,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 100,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 104,
                size: 2,
                opcode: 66,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 106,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 108,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 112,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 114,
                size: 2,
                opcode: 6,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 116,
                size: 2,
                opcode: 66,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 118,
                size: 2,
                opcode: 35,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 120,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 122,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 124,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 126,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 130,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 132,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    168,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 77,
                branch_idx: 3,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 134,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 136,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 140,
                size: 2,
                opcode: 47,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 142,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 144,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    168,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 77,
                branch_idx: 3,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 146,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 148,
                size: 2,
                opcode: 33,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 150,
                size: 2,
                opcode: 36,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 152,
                size: 2,
                opcode: 42,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 154,
                size: 2,
                opcode: 44,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 156,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    168,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 77,
                branch_idx: 3,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 158,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 160,
                size: 2,
                opcode: 17,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 162,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 164,
                size: 2,
                opcode: 54,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 166,
                size: 2,
                opcode: 67,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 168,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    60,
                    65,
                    71,
                ],
                branch_idx: 3,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 170,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    200,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 92,
                branch_idx: 4,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 172,
                size: 2,
                opcode: 35,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 174,
                size: 2,
                opcode: 25,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 176,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    200,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 92,
                branch_idx: 4,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 178,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 180,
                size: 2,
                opcode: 37,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 182,
                size: 2,
                opcode: 42,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 184,
                size: 2,
                opcode: 44,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 186,
                size: 2,
                opcode: 15,
                branch_dest: Some(
                    200,
                ),
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: Some(
            InstructionBranchTo {
                ins_idx: 92,
                branch_idx: 4,
            },
        ),
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 188,
                size: 2,
                opcode: 32,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 190,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 192,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 194,
                size: 2,
                opcode: 46,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 196,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 200,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    78,
                    81,
                    86,
                ],
                branch_idx: 4,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 202,
                size: 2,
                opcode: 35,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 204,
                size: 2,
                opcode: 34,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 206,
                size: 2,
                opcode: 4,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 208,
                size: 4,
                opcode: 19,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 212,
                size: 2,
                opcode: 7,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: Some(
            InstructionBranchFrom {
                ins_idx: [
                    9,
                    26,
                    32,
                ],
                branch_idx: 0,
            },
        ),
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 214,
                size: 2,
                opcode: 55,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 216,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 220,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 224,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 228,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 232,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 236,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
    InstructionDiffRow {
        ins_ref: Some(
            InstructionRef {
                address: 240,
                size: 4,
                opcode: 65534,
                branch_dest: None,
            },
        ),
        kind: None,
        branch_from: None,
        branch_to: None,
        arg_diff: [],
    },
]
